### ISSUE TITLE
Question: re. embedded subtitles

### ISSUE THREAD

**Issue by preacher65** [mdhiggins/sickbeard_mp4_automator#1446]
Hi, would appreciate advice from someone on the correct settings for subtitle handling, as I can't quite figure it out from the wiki.

What would be the correct settings for the following scenarios?

1) Extract all embedded subtitles, and save them as SRT files.

2) Extract embedded subtitles in codec A and save as SRT, but keep as embedded any subtitles using codec B.

Is the second option, extract A but keep B, possible?

Actually option 2 may be the wrong approach. I want to extract subs for the whole episode, if found, and convert to SRT. But if there are subs for a few mins of occasional foreign language dialog, I want those to be left behind. Should I be using disposition settings instead?

TIA

**Comment by mdhiggins** [comment]
2nd option is not really possible unless you specifically wanted to keep something like image based subtitles but not text based subtitles, specific delineation of embed vs extract by codec is not currently possible

Only non-image based subtitles can be extracted SRT (image based cannot be converted to text from SRT) though they can be extracted to other external formats (PGS)

`embed-subs = False`
`codec = srt`

For image based
`embed-image-subs = False`
`codec = pgs`

That should cover most common scenarios, extraction of image based subtitles may require you to add additional codecs to that list depending on what you're encountering on your source files as FFMPEG has limited conversion options for image based subtitles even between image based formats

**Comment by mdhiggins** [comment]
To address your edit

Not really a way to do that but the disposition settings may be helpful there to some degree. Few mins of foreign dialog subtitles are typically flagged as `forced` subtitles since they are designed to supplement a native language. That being said there's no specific options regarding to divide them up on that criteria. That being said they should be labeled appropriately when extracted and applied by your player (plex supports this)

**Comment by preacher65** [comment]
Thanks - will test against one of the offending files. Although I think the issue may lie in the fact the foreign dialogue subs aren't set to forced for these particular files.

I am using Plex, and don't usually have a problem, so this may jut be an edge case. (I think it's happened once or twice before, but this is probably the first time I've tried to rectify it.) I'll also try to find a file with foreign dialogue that Plex handled fine, and compare. Thanks again.

**Comment by preacher65** [comment]
The file in question had an embedded vobsub subtitle track, and as far as I can tell ffmpeg doesn't support vobsub? Which would explain why the embedded track isn't being extracted or converted, just dropped?

If so, it's certainly no problem - this is an odd case and having redownloaded new subs for the file in question, the foreign dialogue subs are included in the new SRT file. If there's a simple way to get it to handle vobsub, great, but otherwise it's too much of an edge case to worry about.

**Comment by mdhiggins** [comment]
I think you can try adding `vobsub` to your image based codecs and see if its able to extract it

**Comment by preacher65** [comment]
`codec = vobsub` gives `Unknown encoder 'vobsub'`


This is the info on the subtitle track:
```
    "subtitle": [
        {
            "index": 2,
            "codec": "dvd_subtitle",
            "disposition": "+default-dub-original-comment-lyrics-karaoke-forced-hearing_impaired-visual_impaired",
            "language": "eng"
        }
    ],
```

ffmpeg seems to want a codec of either `dvdsub` or `dvd_subtitle` for IDX+SUB format subtitles, but it's not extracting them. I've read in #1208 that you're trying an encode to text-based, output to null, and if that fails assume image based. Which explains the first few lines of the following, but I can't work out why it's not trying to extract anything.


```
2021-06-06 13:28:22 - resources.mediaprocessor - INFO - Reading subtitle streams.
2021-06-06 13:28:22 - resources.mediaprocessor - DEBUG - isImageBasedSubtitle FFmpeg command:
2021-06-06 13:28:22 - resources.mediaprocessor - DEBUG - /usr/local/bin/ffmpeg -i /tv/Mr. Robot/Season 02/Mr. Robot (2015) - S02E04 - eps2.2_init_1.asec [Bluray-1080p][10bit][h265][AC3 5.1]-d3g.mkv -c:s:0 srt -map 0:2 -metadata:s:s:0 title= -metadata:s:s:0 handler_name= -metadata:s:s:0 language=und -disposition:s:0 -default-dub-original-comment-lyrics-karaoke-forced-hearing_impaired-visual_impaired-captions -f srt -t 00:00:01 -f null -
2021-06-06 13:28:22 - resources.mediaprocessor - DEBUG - 
2021-06-06 13:28:22 - resources.mediaprocessor - INFO - Image-based subtitle detected for stream 2 - dvd_subtitle eng.
2021-06-06 13:28:22 - resources.mediaprocessor - INFO - Scanned for external subtitles and found 0 results in your approved languages.
2021-06-06 13:28:22 - resources.mediaprocessor - DEBUG - Subtitle output is empty or no default subtitle language is set, will not pass over subtitle output to set a default stream.
```

Not sure if I've got incorrect settings in the autoprocess.ini, or have just misunderstood something fundamental.
```
[Subtitle]
codec = srt
codec-image-based = dvdsub
languages = 
default-language = eng
first-stream-of-language = False
encoding = 
burn-subtitles = False
burn-dispositions = 
embed-subs = False
embed-image-subs = False
embed-only-internal-subs = False
filename-dispositions = forced
ignore-embedded-subs = False
ignored-dispositions = 
unique-dispositions = False
attachment-codec = 
```
If you have any suggestions, they'd be welcome, but I've been able to find external subs with the translated foreign dialogue included so it's not a big issue. Just curious as to whether it can be made to work.

Just to note, if I manually extract the IDX+SUB files and drop them in the relevant folder, Plex plays them fine. Just not sure if ffmpeg can extract them.

**Comment by preacher65** [comment]
Just as test, I used ffmpeg to extract just the vobsub stream into a .mkv file, renamed it to .mks, dropped it in the relevant media folder and Plex found and displayed it.

Which all seems like a lot of hassle to retain subtitles from the source when I've been able to find them online...

**Comment by mdhiggins** [comment]
What command did you used to extract the sub?

**Comment by preacher65** [comment]
`ffmpeg.exe -i sourcefile.mkv -map 0:s:0 -scodec dvdsub outfile.mkv`
Although I don't really know the options that well so it was a bit trial and error. Have since discovered the `-f` switch so now trying
`ffmpeg.exe -i sourcefile.mkv -map 0:s:0 -scodec dvdsub -f matroska outfile.mks`
and
`ffmpeg.exe -i sourcefile.mkv -map 0:s:0 -codec copy -f matroska outfile.mks`
which both seem to work.

**Comment by mdhiggins** [comment]
Just made a small update that should make this work

Give it a try

**Comment by preacher65** [comment]
Thanks. Still getting
```
2021-06-06 17:49:31 - resources.mediaprocessor - INFO - Reading subtitle streams.
2021-06-06 17:49:31 - resources.mediaprocessor - DEBUG - isImageBasedSubtitle FFmpeg command:
2021-06-06 17:49:31 - resources.mediaprocessor - DEBUG - /usr/local/bin/ffmpeg -i /tv/Mr. Robot/Season 02/Mr. Robot (2015) - S02E04 - eps2.2_init_1.asec [Bluray-1080p][10bit][h265][AC3 5.1]-d3g.mkv -c:s:0 srt -map 0:2 -metadata:s:s:0 title= -metadata:s:s:0 handler_name= -metadata:s:s:0 language=und -disposition:s:0 -default-dub-original-comment-lyrics-karaoke-forced-hearing_impaired-visual_impaired-captions -f srt -t 00:00:01 -f null -
2021-06-06 17:49:31 - resources.mediaprocessor - DEBUG - 
2021-06-06 17:49:31 - resources.mediaprocessor - INFO - Image-based subtitle detected for stream 2 - dvd_subtitle eng.
2021-06-06 17:49:31 - resources.mediaprocessor - INFO - Scanned for external subtitles and found 0 results in your approved languages.
2021-06-06 17:49:31 - resources.mediaprocessor - DEBUG - Subtitle output is empty or no default subtitle language is set, will not pass over subtitle output to set a default stream.
```
Checked that extensions.py had updated and it has. Am I missing something obvious from autoprocess.ini?
```
[Subtitle]
codec = srt
codec-image-based = dvdsub,dvd_subtitle
languages = eng
default-language = eng
first-stream-of-language = False
encoding = 
burn-subtitles = False
burn-dispositions = 
embed-subs = False
embed-image-subs = False
embed-only-internal-subs = False
filename-dispositions = default,forced
ignore-embedded-subs = False
ignored-dispositions = 
unique-dispositions = False
attachment-codec = 
```

**Comment by mdhiggins** [comment]
Can you perhaps send me the file you're testing

**Comment by preacher65** [comment]
Also just to say "doh" regarding my comment about the image subs causing transcoding in Plex, which I assume is as expected as plex has to burn the subs into the video stream. More surprising that the new Plex HTPC player doesn't cause transcoding, if I'm understanding this correctly.

**Comment by mdhiggins** [comment]
Yeah there's a few players that can do image based subs without trans coding. I believe the nvidia shield is another

**Comment by preacher65** [comment]
File is on its way - link via email to the address on your github. It'll take at least 1/2 hour to upload though.

**Comment by preacher65** [comment]
Just to add that from my POV this is an interesting exercise, but in no way is it important - the original problem has been overcome just by finding an srt with the translated dialogue included. So if you're interested by all means take a look, but please don't spend much time on it.

**Comment by mdhiggins** [comment]
Still good to make sure that the script is as robust as possible and can handle some of these edge cases. Helps me learn things too

9fc00753450c40a534154c448715ff036c135f9f

That seems to allow the subtitles to be extracted appropriately now using `codec-image-based = dvd_subtitle`

**Comment by preacher65** [comment]
Hmm. Sort of works. ;)
Restarted my sonarr-sma docker to update the script (SMA_UPDATE set to TRUE). Connected to shell and checked the formats.py had updated. Imported file in Sonarr and it still gives the following in the log:
```
2021-06-07 00:46:36 - resources.mediaprocessor - INFO - Image-based subtitle detected for stream 2 - dvd_subtitle eng.
2021-06-07 00:46:36 - resources.mediaprocessor - INFO - Scanned for external subtitles and found 0 results in your approved languages.
2021-06-07 00:46:36 - resources.mediaprocessor - DEBUG - Subtitle output is empty or no default subtitle language is set, will not pass over subtitle output to set a default stream.
```
Installed SMA on my Windows machine, copied the autoprocess.ini from the docker container and changed the path to ffmpeg, then ran a manual conversion and it worked.
Tried to run manual.py from the docker shell and it complains that guessit isn't installed, not sure if it's possible to run manual conversions from the shell on the docker?
My full autoprocess.ini is [here](https://pastebin.com/e9znj5BK) in case I've done something stupid. At this point it kind of feels like I may have. :)
I'm using the build tag to make use of VAAPI so will kick off a rebuild of the container and try again.

**Comment by mdhiggins** [comment]
To run manual in the container you need to use the python virtual environment that has all the dependencies

`/usr/local/sma/venv/bin/python3 /usr/local/sma/manual.py`

That part of the logs you linked is probably a little too early to show the rip process, did a mks file get produced?

**Comment by preacher65** [comment]
You're right (natch), I was looking too early in the logs.
```
2021-06-06 17:49:31 - resources.mediaprocessor - ERROR - Unable to create external mks subtitle file for stream 2, may be an incompatible format.
```
has been replaced with 
```
2021-06-07 00:46:37 - resources.mediaprocessor - INFO - /filename.eng.default.mks created.
```
But it wasn't present in the folder.
Just re-importing after rebuilding the container, and watching the output. The mks file is there, I'll see if something is deleting it...

So I watched and the following happened:

- conversion started
- filename.default.eng.mks was created
- filename.eng.srt was downloaded by Bazarr
- conversion completed
- Sonarr imported the file
- both the .srt and the .mks were deleted.

And sure enough. I found the subtitle files in Sonarr's recycle bin folder, and can see it deleting them in the Sonarr logs.

I'm guessing it's something to do with "Import Extra Files" in Sonarr's Media Management. Will have to test some more tomorrow. But the subtitle is now being created. Thanks!
